core/src/apps/base.py

```diff
@@ -183,6 +183,64 @@ async def handle_DoPreauthorized(
     return await handler(ctx, req, authorization.get())  # type: ignore [Expected 2 positional arguments]
 
 
+async def handle_UnlockPath(ctx: wire.Context, msg: UnlockPath) -> protobuf.MessageType:
+    from trezor.crypto import hmac
+    from trezor.messages import Cancel, UnlockedPathRequest
+    from trezor.ui.layouts import confirm_action
+    from apps.common.paths import SLIP25_PURPOSE
+    from apps.common.seed import Slip21Node, get_seed
+    from apps.common.writers import write_uint32_le
+
+    _KEYCHAIN_MAC_KEY_PATH = [b"TREZOR", b"Keychain MAC key"]
+
+    # UnlockPath is relevant only for SLIP-25 paths.
+    # Note: Currently we only allow unlocking the entire SLIP-25 purpose subtree instead of
+    # per-coin or per-account unlocking in order to avoid UI complexity.
+    if len(msg.address_n) != 1 or msg.address_n[0] != SLIP25_PURPOSE:
+        raise wire.DataError("Invalid path")
+
+    seed = await get_seed(ctx)
+    node = Slip21Node(seed)
+    node.derive_path(_KEYCHAIN_MAC_KEY_PATH)
+    mac = utils.HashWriter(hmac(hmac.SHA256, node.key()))
+    for i in msg.address_n:
+        write_uint32_le(mac, i)
+    expected_mac = mac.get_digest()
+
+    # Require confirmation to access SLIP25 paths unless already authorized.
+    if msg.mac:
+        if len(msg.mac) != len(expected_mac) or not utils.consteq(
+            expected_mac, msg.mac
+        ):
+            raise wire.DataError("Invalid MAC")
+    else:
+        await confirm_action(
+            ctx,
+            "confirm_coinjoin_access",
+            title="CoinJoin account",
+            description="Do you want to allow access to CoinJoin accounts?",
+        )
+
+    wire_types = (
+        MessageType.Cancel,
+        MessageType.GetAddress,
+        MessageType.GetPublicKey,
+        MessageType.SignTx,
+    )
+    req = await ctx.call_any(UnlockedPathRequest(mac=expected_mac), *wire_types)
+    if Cancel.is_type_of(req):
```

Comment:
do we need explicit handling of Cancel? if you leave it out, receiving Cancel will raise `UnexpectedMessage` out of this handler, and invoke the `Cancel` handler, to the same effect istm

Reply:
The implicit handling of `Cancel` seems to work just fine. https://github.com/trezor/trezor-firmware/pull/2289/commits/aafaadabaedac2e0f35e98a07985fb782bbd19b0